src/Connection.php

```diff
@@ -173,21 +173,21 @@ public function queryStream($sql, $params = array())
         return $stream;
     }
 
-    /**
-     * {@inheritdoc}
-     */
-    public function ping($callback)
+    public function ping()
     {
-        if (!is_callable($callback)) {
-            throw new \InvalidArgumentException('Callback is not a valid callable');
+        $deferred = new Deferred();
+        try {
+            $this->_doCommand(new PingCommand($this))
+                ->on('error', function ($reason) use ($deferred) {
+                    $deferred->reject($reason);
+                })
+                ->on('success', function () use ($deferred) {
+                    $deferred->resolve(true);
+                });
+        } catch (\Exception $e) {
+            $deferred->reject($e);
         }
-        $this->_doCommand(new PingCommand($this))
-            ->on('error', function ($reason) use ($callback) {
-                $callback($reason, $this);
-            })
-            ->on('success', function () use ($callback) {
-                $callback(null, $this);
-            });
+        return $deferred->promise();
```

Comment:
You could be optimized by using a `Promise`directly.

```php
return new Promise(function ($resolve, $reject) {
    $this->_doCommand(new PingCommand($this))
        ->on('error', function ($reason) use ($reject) {
            $reject($reason);
        })
        ->on('success', function () use ($resolve) {
            $resolve(true);
        });
});
```

Reply:
Good catch, updated as per your suggestion! :+1: